### ISSUE TITLE
pod with custom service account unable to accept any incoming traffic (mtls)

### ISSUE THREAD

**Issue by Stono** [istio/istio#2774]
Hey all, 
We've been looking at this for days now and getting absolutely nowhere, so would really appreciate some support.  We've been using istio for a while now without any issue, but the is the first time we've brought into scope service accounts for some pods.

We have a pod, which is running under a custom service account called `platform-helm-deployer-account`.  There is (correctly) a certificate generated for the service account:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            cb:87:45:f9:d6:0d:a2:bd:88:c7:f5:c8:7f:62:fd:0a
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: O=k8s.cluster.local
        Validity
            Not Before: Jan 23 19:51:06 2018 GMT
            Not After : Jan 23 20:51:06 2018 GMT
        Subject: O=
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
            RSA Public Key: (2048 bit)
                Modulus (2048 bit):
                     ....
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Client Authentication, TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name:
                URI:spiffe://cluster.local/ns/default/sa/platform-helm-deployer-account
```

So the spiffe looks fine, however - when attempting to connect to the service even from the pod itself, we get an error from envoy:

```
[root@platform-helm-deployer-76d6bc5fb4-mkn7m app]# curl -kvL platform-helm-deployer-service
* About to connect() to platform-helm-deployer-service port 80 (#0)
*   Trying 10.178.5.250...
* Connected to platform-helm-deployer-service (10.178.5.250) port 80 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: platform-helm-deployer-service
> Accept: */*
>
< HTTP/1.1 503 Service Unavailable
< content-length: 57
< content-type: text/plain
< date: Tue, 23 Jan 2018 20:26:10 GMT
< server: envoy
<
* Connection #0 to host platform-helm-deployer-service left intact
```

We don't see much in istio-proxy logs:
```
[libprotobuf INFO external/mixerclient_git/src/check_cache.cc:155] Add a new Referenced for check cache: Absence-keys: source.uid, Exact-keys: context.protocol, destination.ip, destination.service, destination.uid, source.ip,
[libprotobuf INFO external/mixerclient_git/src/check_cache.cc:155] Add a new Referenced for check cache: Absence-keys: Exact-keys: context.protocol, destination.service,
[2018-01-23T20:33:01.827Z] "GET / HTTP/1.1" 503 UF 0 57 24 - "-" "curl/7.29.0" "08a385f0-372a-9d55-abd1-1998a0b08c9a" "platform-helm-deployer-service" "10.177.0.22:8080"
```

I can't really see anything in the mixer logs:
```
I0123 20:22:07.636292       1 grpcServer.go:137] Dispatching Preprocess Check
I0123 20:22:07.636352       1 dispatcher.go:165] Resolved (TEMPLATE_VARIETY_ATTRIBUTE_GENERATOR) 1 actions
I0123 20:22:07.636517       1 kubernetesenv.go:211] handler.kubernetesenv.istio-system:could not find pod for (uid: kubernetes://platform-helm-deployer-76d6bc5fb4-mkn7m.default, key: default/platform-helm-deployer-76d6bc5fb4-mkn7m)
I0123 20:22:07.636547       1 kubernetesenv.go:211] handler.kubernetesenv.istio-system:could not find pod for (uid: 10.177.0.20, key: 10.177.0.20)
I0123 20:22:07.636761       1 grpcServer.go:156] Preprocess Check returned with: OK
I0123 20:22:07.636776       1 grpcServer.go:159] Dispatching to main adapters after running processors
I0123 20:22:07.636854       1 grpcServer.go:160] Attribute Bag:
context.protocol              : http
destination.ip                : [0 0 0 0 0 0 0 0 0 0 255 255 10 177 0 20]
destination.service           : platform-helm-deployer-service.default.svc.cluster.local
destination.uid               : kubernetes://platform-helm-deployer-76d6bc5fb4-mkn7m.default
request.headers               : {request.headers map[x-b3-sampled:1 x-b3-spanid:fee8e398f3cacd0c x-b3-traceid:fee8e398f3cacd0c x-request-id:119e4431-76a9-9e4d-83ec-1bde91330950 accept:*/* user-agent:curl/7.29.0 :authority:platform-helm-deployer-service :method:GET :path:/ x-envoy-decorator-operation:default-route x-ot-span-context:fee8e398f3cacd0c;fee8e398f3cacd0c;0000000000000000 x-forwarded-proto:http] 0xc420b6a4e0}
request.host                  : platform-helm-deployer-service
request.method                : GET
request.path                  : /
request.scheme                : http
request.time                  : 2018-01-23 20:22:07.730018028 +0000 UTC
request.useragent             : curl/7.29.0
source.ip                     : [10 177 0 20]
source.port                   : 46446
---
destination.labels            : map[pod-template-hash:3282671960 version:6 app:platform-helm-deployer]
destination.namespace         : default
destination.serviceAccount    : platform-helm-deployer-account
source.labels                 : map[version:6 app:platform-helm-deployer pod-template-hash:3282671960]
source.namespace              : default
source.service                : platform-helm-deployer.default.svc.cluster.local
source.serviceAccount         : platform-helm-deployer-account
I0123 20:22:07.636867       1 grpcServer.go:163] Dispatching Check
I0123 20:22:07.636884       1 dispatcher.go:165] Resolved (TEMPLATE_VARIETY_CHECK) 0 actions
I0123 20:22:07.636901       1 grpcServer.go:186] Check returned with ok
I0123 20:22:07.636907       1 grpcServer.go:243] Dispatching Quota: RequestCount
I0123 20:22:07.636922       1 dispatcher.go:165] Resolved (TEMPLATE_VARIETY_QUOTA) 0 actions
```

What I find very interesting however is that curling using the certificates on the istio-proxy work fine:
```
istio-proxy@platform-helm-deployer-bc445fdc8-jj7xd:/$ curl https://platform-helm-builder-service -v --key /etc/certs/key.pem --cert /etc/certs/cert-chain.pem --cacert /etc/certs/root-cert.pem -k > /dev/null
* Rebuilt URL to: https://platform-helm-builder-service/
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 10.178.6.42...
* Connected to platform-helm-builder-service (10.178.6.42) port 443 (#0)
* found 1 certificates in /etc/certs/root-cert.pem
* found 0 certificates in /etc/ssl/certs
* ALPN, offering http/1.1
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256
*        server certificate verification SKIPPED
*        server certificate status verification SKIPPED
* error fetching CN from cert:The requested data were not available.
*        common name:  (does not match 'platform-helm-builder-service')
*        server certificate expiration date OK
*        server certificate activation date OK
*        certificate public key: RSA
*        certificate version: #3
*        subject: O=#1300
*        start date: Tue, 23 Jan 2018 19:43:13 GMT
*        expire date: Tue, 23 Jan 2018 20:43:13 GMT
*        issuer: O=k8s.cluster.local
*        compression: NULL
* ALPN, server did not agree to a protocol
> GET / HTTP/1.1
> Host: platform-helm-builder-service
> User-Agent: curl/7.47.0
> Accept: */*
>
< HTTP/1.1 200 OK
```

So it appears to be something to do with egress traffic leaving a pod to that service not being intercepted / mtls'd ??

look forward to your support!

**Comment by Stono** [comment]
Sorry I should have said this is istio `0.4.0`, on GKE `1.8.6`

**Comment by Stono** [comment]
OK literally 10 minutes after writing this post, after looking for day - I got to the bottom of it.
The container was running with the wrong secret mounted, instead of `istio.service-account-name` it was mounting `istio.default`.  This was due to an error in our helm chart (which actually manually injects istio).  So, user, not istio error.

However, it would be really great to have some pointers around debugging such issues however, as i've sunk so much time into this and didn't really know where to look for more information!
